pkg/backend/filestate/backend.go

```diff
@@ -233,19 +237,117 @@ func New(ctx context.Context, d diag.Sink, originalURL string, project *workspac
 	}
 	backend.currentProject.Store(project)
 
+	// projectMode tracks whether the current state supports project-scoped stacks.
+	// Historically, the filestate backend did not support this.
+	// To avoid breaking old stacks, we use legacy mode for existing states.
+	// We use project mode only if one of the following is true:
+	//
+	//  - The state has a .pulumi/Pulumi.yaml file (the 'meta' file)
+	//    and the version is 1 or greater.
+	//  - The state is entirely new so there's no risk of breaking old stacks.
+	//
+	// All actual logic of project mode vs legacy mode is handled by the referenceStore.
+	// This boolean just helps us warn users about unmigrated stacks.
+	var projectMode bool
 	switch v := pulumiState.Version; v {
 	case 0:
 		backend.store = newLegacyReferenceStore(wbucket)
 	case 1:
 		backend.store = newProjectReferenceStore(wbucket, backend.currentProject.Load)
+		projectMode = true
 	default:
 		return nil, fmt.Errorf(
 			"state store unsupported: 'Pulumi.yaml' version (%d) is not supported "+
 				"by this version of the Pulumi CLI", v)
 	}
+
+	// If we're not in project mode, we're done.
+	if !projectMode {
+		return backend, nil
+	}
+	// Otherwise, warn about any old stack files.
+	// This is possible if a user creates a new stack with a new CLI,
+	// or migrates it to project mode with `pulumi state upgrade`,
+	// but someone else interacts with the same state with an old CLI.
+
+	refs, err := newLegacyReferenceStore(wbucket).ListReferences()
+	if err != nil {
+		// If there's an error listing don't fail, just don't print the warnings
+		return backend, nil
+	}
+
+	for _, ref := range refs {
+		d.Warningf(&diag.Diag{
+			Message: "Found legacy stack file '%s', you should run 'pulumi state upgrade'",
+		}, ref.Name())
+	}
+
 	return backend, nil
 }
 
+func (b *localBackend) Upgrade(ctx context.Context) error {
```

Comment:
This method is clearly fine as part of this PR, just to clarify my comment above